motoko/icrc2-swap/src/swap/main.mo

```diff
@@ -0,0 +1,294 @@
+import TrieMap "mo:base/TrieMap";
+
+// Import our ICRC type definitions
+import ICRC "./ICRC";
+
+// The swap canister is the main backend canister for this example. To simplify
+// this example we configure the swap canister with the two tokens it will be
+// swapping.
+shared(init_msg) actor class Swap(init_args: {
+  token_a: Principal;
+  token_b: Principal;
+}) = this {
+
+  // Track the deposited per-user balances for token A and token B
+  private var balancesA = TrieMap.TrieMap<Principal, Nat>(Principal.equal, Principal.hash);
+  private var balancesB = TrieMap.TrieMap<Principal, Nat>(Principal.equal, Principal.hash);
+
+  // Because TrieMaps are not directly storeable in stable memory we need a
+  // location to store the data during canister upgrades.
+  private stable var stableBalancesA : ?[(Principal, Nat)] = null;
+  private stable var stableBalancesB : ?[(Principal, Nat)] = null;
+
+  // balances is a simple getter to check the balances of all users, to make debugging easier.
+  public query func balances(): async ([(Principal, Nat)], [(Principal, Nat)]) {
+    (Iter.toArray(balancesA.entries()), Iter.toArray(balancesB.entries()))
+  };
+
+  public type DepositArgs = {
+    spender_subaccount: ?Blob;
+    token: Principal;
+    from: ICRC.Account;
+    amount: Nat;
+    fee: ?Nat;
+    memo: ?Blob;
+    created_at_time: ?Nat64;
+  };
+
+  public type DepositError = {
+    #TransferFromError: ICRC.TransferFromError;
+  };
+
+  // Accept deposits
+  // - Accept TokenA, and TokenB
+  // - user approves transfer: `token_a.icrc2_approve({ spender=swap_canister; amount=amount; ... })`
+  // - user deposits their token: `swap_canister.deposit({ token=token_a; amount=amount; ... })`
+  // - These deposit handlers show how to safely accept and register deposits of an ICRC-2 token. 
+  public shared(msg) func deposit(args: DepositArgs): async Result.Result<Nat, DepositError> {
+    let token: ICRC.Actor = actor(Principal.toText(args.token));
+    let balances = which_balances(args.token);
+
+    // Load the fee from the token here. The user can pass a null fee, which
+    // means "use the default". So we need to look up the default in order to
+    // correctly deduct it from their balance.
+    let fee = switch (args.fee) {
+      case (?f) { f };
+      case (null) { await token.icrc1_fee() };
+    };
+
+    // Perform the transfer, to capture the tokens.
+    let transfer_result = await token.icrc2_transfer_from({
+      spender_subaccount = args.spender_subaccount;
+      from = args.from;
+      to = { owner = Principal.fromActor(this); subaccount = null };
+      amount = args.amount;
+      fee = ?fee;
+      memo = args.memo;
+      created_at_time = args.created_at_time;
+    });
+
+    // Check that the transfer was successful.
+    let block_height = switch (transfer_result) {
+      case (#Ok(block_height)) { block_height };
+      case (#Err(err)) {
+        // Transfer failed. There's no cleanup for us to do, so we can just
```

Comment:
```suggestion
        // Transfer failed. There's no cleanup for us to do since no state has changed, so we can just
```
Since this is aimed at beginners I think it's worth being very explicit